### PR TITLE
[Companion] Bound Election and Staking by MaxActiveValidators

### PR DESCRIPTION
**Pull request by Ank4n**
Companion for https://github.com/paritytech/substrate/pull/12436

**Changes:**
- Election provider pallet (election-provider-multi-phase) is bounded by `MaxWinners` it can elect.
- Staking pallet is bounded by `MaxWinners` which it derives from pallet election. This `MaxWinners`  is essentially the `MaxActiveValidators` for the staking pallet.  
- Replace unbounded and bounded on-chain election implementation into a single impl `OnChainExecution` which is always bounded.

### CODE REVIEW

File: runtime/kusama/src/lib.rs [hunk 1]
```diff
@@ -428,6 +429,9 @@ impl onchain::Config for OnChainSeqPhragmen {
 	type Solver = SequentialPhragmen<AccountId, runtime_common::elections::OnChainAccuracy>;
 	type DataProvider = Staking;
 	type WeightInfo = weights::frame_election_provider_support::WeightInfo<Runtime>;
+	type MaxWinners = MaxActiveValidators;
+	type VotersBound = MaxElectingVoters;
```
Comment (kianenigma) [inline runtime/kusama/src/lib.rs:433]:
```suggestion
	type VotersBound = MaxElectingVoters / 4;
```

File: runtime/kusama/src/lib.rs [hunk 2]
```diff
@@ -428,6 +429,9 @@ impl onchain::Config for OnChainSeqPhragmen {
...
+	type TargetsBound = MaxElectableTargets;
```
Comment (kianenigma) [inline runtime/kusama/src/lib.rs:434]:
```suggestion
	type TargetsBound = MaxElectableTargets / 4;
```

File: runtime/polkadot/src/lib.rs [hunk 3]
```diff
@@ -405,6 +405,7 @@ parameter_types! {
 	/// ... and all of the validators as electable targets. Whilst this is the case, we cannot and
 	/// shall not increase the size of the validator intentions.
 	pub const MaxElectableTargets: u16 = u16::MAX;
+	pub const MaxActiveValidators: u32 = 1000;
```
[the remark this answers is not in the document]
Comment (kianenigma) [reply runtime/polkadot/src/lib.rs:408]: Task: what if they are equal. we need to make sure if they are equal, nothing breaks.

Comment (kianenigma) [inline runtime/polkadot/src/lib.rs:408]:
```suggestion
	pub const MaxActiveValidators: u32 = 1200;
```

File: runtime/test-runtime/src/lib.rs [hunk 4]
```diff
@@ -319,6 +319,9 @@ parameter_types! {
 	pub storage MaxNominatorRewardedPerValidator: u32 = 64;
 	pub storage OffendingValidatorsThreshold: Perbill = Perbill::from_percent(17);
 	pub const MaxAuthorities: u32 = 100_000;
+	pub OnChainMaxWinners: u32 = 100;
```
Comment (kianenigma) [inline runtime/test-runtime/src/lib.rs:322]:
```suggestion
	pub OnChainMaxWinners: u32 = 1000;
```

File: utils/staking-miner/src/emergency_solution.rs [hunk 5]
```diff
@@ -36,17 +36,17 @@ macro_rules! emergency_solution_cmd_for { ($runtime:ident) => { paste::paste! {
 
 			log::info!(target: LOG_TARGET, "mined solution with {:?}", &raw_solution.score);
 
-			let mut ready_solution = EPM::Pallet::<Runtime>::feasibility_check(raw_solution, EPM::ElectionCompute::Signed)?;
-
+			let ready_solution = EPM::Pallet::<Runtime>::feasibility_check(raw_solution, EPM::ElectionCompute::Signed)?;
+			let mut supports = ready_solution.supports.clone().into_inner();
```
Comment (kianenigma) [inline utils/staking-miner/src/emergency_solution.rs:40]:
```suggestion
			let mut supports = ready_solution.supports.into_inner();
```

**Review by kianenigma — APPROVED** [review]

**Review by ruseinov — APPROVED** [review]

**Review by gpestana — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Pull request merged**